### PR TITLE
migrate gram muon to pytorch (#183818)

### PR DESCRIPTION
**Pull request by tangyuhao**
Summary:

## Context

migrate Gram Muon support to Pytorch lib. 

The Gram Muon algorithm is from https://github.com/Dao-AILab/gram-newton-schulz/tree/main. from our tests, it achieve same NE as vanilla muon but better speed (up to 2x). 

Besides, we added some more optimization like cuda graph and batch support which is very usefully to reduce kernel overhead for many same-shape weights in the model.

## What this diff does

Adds an opt-in Gram Newton-Schulz path to `torch.optim.Muon` while leaving the default behavior unchanged.

- New `use_gram_newton_schulz: bool = False` constructor flag selects the alternate path. With the default `False`, dispatch in `muon()` early-returns into the existing `_single_tensor_muon` with the same kwargs as before — the optimizer is bit-compatible with prior `torch.optim.Muon`.
- New `gram_newton_schulz_config: GramNewtonSchulzConfig | None = None` carries gram-only knobs: per-iteration `(a, b, c)` coefficients, `gram_ns_reset_iterations`, and `use_cuda_graph`. `_parse_gram_newton_schulz_config` validates keys against the `TypedDict` schema and rejects the dict when the flag is off so callers cannot configure CUDA graph on the vanilla path silently.
- `_gram_muon_impl` is the multi-tensor implementation. **Phase 1** fuses momentum updates across all params via `torch._foreach_lerp_`, **Phase 2** groups params by shape and runs the gram orthogonalizer per group (`_gram_newton_schulz`, `_gram_newton_schulz_batched`, with auto-fallback to standard NS for square matrices), and **Phase 3** fuses the parameter update via `torch._foreach_mul_` for the uniform weight-decay scale plus one `torch._foreach_add_` per shape group for the shape-specific adjusted learning rate.
- Optional CUDA graph capture (`gram_newton_schulz_config={"use_cuda_graph": True}`) caches per-shape NS computations on `Muon._cuda_graph_cache` keyed by shape, dtype, device, coefficients, and reset iterations. Square shapes always run eager.

Test Plan:
## Test target

```
buck test fbcode//caffe2/test:optim -- TestMuon
```

Both `fbcode/caffe2/test/optim/test_muon.py` and `xplat/caffe2/test/optim/test_muon.py` are added (kept in sync). The `glob(["optim/*.py"])` in `caffe2/test/BUCK` auto-registers the new file — no `TARGETS` edit required.

## Test classes

| Class | Coverage |
|---|---|
| `TestMuonStandardNS` | Vanilla path on rectangular/square/multi-param shapes, weight decay, nesterov toggle. |
| `TestMuonGramNS` | Gram path on rectangular shapes, square-fallback, multi-step, multi-param. Includes `test_default_is_vanilla` (default kwargs match explicit `use_gram_newton_schulz=False`) and `test_gram_vs_standard_differ` (paths produce different updates for rectangular params). |
| `TestMuonGramConfig` | Validates dict schema: unknown keys rejected, dict-with-flag-off rejected, partial overrides preserve defaults, full overrides respected. |
| `TestMuonConvergence` | End-to-end convergence on a learnable 2-layer linear regression target — both gram and vanilla paths must drop loss by **>= 90%** in 200 steps. |
| `TestMuonCudaGraph` | Eager-vs-graph parity (single param, multi-param same shape, mixed-config groups, tall matrix), multi-step graph correctness. Skipped when CUDA is unavailable. |
| **`TestMuonVanillaRegression`** | **Frozen-value bit-compatibility lock.** Captured at the parent of D104311008 and asserts `atol=0, rtol=0` against the new dispatch wrapper. |

## How the regression baseline was captured

1. `sl goto b39ef94f02^` — hop to parent commit.
2. Wrote a temporary `test_muon_baseline_capture.py` that constructs a fixed-seed `Muon([nn.Parameter(torch.randn(8, 4))], lr=0.02, momentum=0.95, ...)`, runs 5 steps with `torch.randn_like(p)` grads, and dumps `param.tolist()` + `state["momentum_buffer"].tolist()` to JSON for two scenarios (`nesterov=True, weight_decay=0.1` and `nesterov=False, weight_decay=0.0`).
3. Ran `buck test fbcode//caffe2/test:optim -- TestMuonBaselineCapture` to execute the capture against the pre-diff vanilla `Muon`.
4. Read the JSON, removed the temp file, `sl goto b39ef94f02` back to the diff commit.
5. Embedded the captured float32 literals as class constants on `TestMuonVanillaRegression` and replayed the same setup with `use_gram_newton_schulz=False`.

## Result

```
buck test fbcode//caffe2/test:optim -- TestMuonVanillaRegression
Tests finished: Pass 2. Fail 0. Timeout 0. Fatal 0. Skip 0. Omit 0. Infra Failure 0. Build failure 0
```

Both `test_vanilla_nesterov_true_matches_pre_diff_baseline` and `test_vanilla_nesterov_false_matches_pre_diff_baseline` pass with `atol=0, rtol=0`, confirming the new dispatch wrapper preserves vanilla behavior bit-for-bit on CPU.

## Lint

```
arc lint -a fbcode/caffe2/test/optim/test_muon.py xplat/caffe2/test/optim/test_muon.py
ok No lint issues.
```

Differential Revision: D104311008

### CODE REVIEW

File: torch/optim/_muon.py [hunk 1]
```diff
@@ -70,6 +222,41 @@ def _zeropower_via_newtonschulz(
...
+    if len(grad.shape) != 2:
+        raise ValueError("Input tensor gradient must be a 2D matrix")
+
+    X = grad.float()
+
+    if should_transpose := (X.size(0) > X.size(1)):
+        X = X.T
+
+    X /= X.norm() + eps
+    X = X.half()
```
Comment (Skylion007) [inline torch/optim/_muon.py:247]: Nit could the output of the prev norm call just be half? Or does that hurt numeric?

File: torch/optim/_muon.py [hunk 2]
```diff
@@ -132,6 +370,13 @@ def __init__(
                         f"Muon only supports 2D parameters whereas we found a parameter with size: {p.size()}"
                     )
 
+        # CUDA graph cache: keyed by NS config, stores
+        # (graph, static_input_bufs, static_output_bufs).
+        self._cuda_graph_cache: dict[
+            tuple[Any, ...],
```
Comment (Skylion007) [inline torch/optim/_muon.py:376]: Should this be a TypeVarTuple or a properly typed tuple for the key? The key at least always have the same n-arry for the keys right?

File: torch/optim/_muon.py [hunk 3]
```diff
@@ -305,6 +579,181 @@ def step(self, closure=None):
...
+
+        # Capture: static_inputs are read, static_outputs are written.
+        graph = torch.cuda.CUDAGraph()
+        with torch.cuda.graph(graph):
+            static_outputs = compute_fn(static_inputs)
+        cuda_graph_cache[cache_key] = (graph, static_inputs, static_outputs)
+
+    graph, static_inputs, static_outputs = cuda_graph_cache[cache_key]
+    # Copy new data into the static input buffers the graph reads from.
+    for si, t in zip(static_inputs, inputs):
```
Comment (Skylion007) [inline torch/optim/_muon.py:614]: strict=True for zip

File: torch/optim/_muon.py [hunk 4]
```diff
@@ -305,6 +579,181 @@ def step(self, closure=None):
...
+            group_inputs,
+            use_cuda_graph=use_cuda_graph,
+            cuda_graph_cache=cuda_graph_cache,
+            gram_ns_coefficients=gram_ns_coefficients,
+            gram_ns_reset_iterations=gram_ns_reset_iterations,
+            ns_coefficients=ns_coefficients,
+            ns_steps=ns_steps,
+            eps=eps,
+        )
+        for idx, r in zip(indices, group_results):
```
Comment (Skylion007) [inline torch/optim/_muon.py:752]: Same here

File: torch/optim/_muon.py [hunk 5]
```diff
@@ -357,17 +889,47 @@ def muon(
     eps: float,
     adjust_lr_fn: str | None,
     has_complex: bool,
+    use_gram_newton_schulz: bool = False,
```
Comment (Skylion007) [inline torch/optim/_muon.py:892]: Boolean trap? Should be impl enum maybe?

**Comment by tangyuhao** [comment]
internally discussed, will split this request to several changes so close it
